contentcuration/contentcuration/static/js/edit_channel/sharedComponents/Uploader.vue

```diff
@@ -0,0 +1,276 @@
+        // Get unsupported file types
+        let partition = _.partition(files, f =>
+          _.contains(this.acceptedExtensions, _.last(f.name.split('.')).toLowerCase())
+        );
+        files = partition[0];
+        this.unsupportedFiles = partition[1];
+
+        // Get files that exceed the max file size
+        partition = _.partition(files, f => f.size < MAX_FILE_SIZE);
+        files = partition[0];
+        this.tooLargeFiles = partition[1];
+
+        // Get the total file size
+        this.totalUploadSize = _.reduce(files, (sum, f) => sum + f.size, 0);
+        return files;
+      },
+      handleFiles(files) {
+        if (!this.readonly) {
+          let newFiles = [];
+          files = this.allowMultiple ? files : [files[0]];
+          files = this.validateFiles(files);
+
+          // Show errors if relevant
+          if (this.totalUploadSize > this.availableSpace) {
+            this.$refs.storageexceeded.prompt();
+            return;
+          } else if (this.unsupportedFiles.length) {
+            this.$refs.unsupportedfiles.prompt();
+          } else if (this.tooLargeFiles.length) {
+            this.$refs.toolargefiles.prompt();
+          }
+
+          [...files].forEach(uploadedFile => {
+            let fileID = String(Math.random()).slice(2);
+            this.addFile({ id: fileID, file: uploadedFile, preset: this.presetID });
+            let file = this.getFile(fileID);
+            newFiles.push(file);
+
+            this.uploadFile({ file: uploadedFile, id: fileID })
+              .then(filepath => {
+                this.$emit('uploaded', filepath);
+              })
+              .catch(error => {
+                this.setError(fileID, error);
+              });
+          });
+          if (newFiles.length) {
+            this.$emit('uploading', newFiles);
+          }
+        }
+      },
+    },
+    $trs: {
+      [fileErrors.NO_STORAGE]: 'Out of storage',
+      [fileErrors.WRONG_TYPE]: 'Invalid file type (must be {filetypes})',
+      [fileErrors.TOO_LARGE]: 'File too large. Must be under {size}',
+      [fileErrors.UPLOAD_FAILED]: 'Upload failed',
+      unsupportedFilesHeader: 'Unsupported files',
+      unsupportedFilesText:
+        '{count, plural,\n =1 {File}\n other {# files}} will not be uploaded.\n' +
+        ' {extensionCount, plural,\n =1 {Accepted file type is}\n other {Accepted file types are}} {extensions}',
+      listDelimiter: ', ',
+      noStorageHeader: 'Not enough space',
+      uploadSize: 'Upload is too large: {size}',
+      remainingStorage: 'Remaining storage: {size}',
+      tooLargeFilesHeader: 'Max file size exceeded',
+      maxFileSizeText:
+        '{count, plural,\n =1 {File}\n other {# files}} will not be uploaded. File size must be under {size}',
+    },
+  };
+
+</script>
+
+<style lang="less" scoped>
+
+  .storage-alert {
+    font-size: 12pt;
+    .storage-usage {
+      margin-top: -5px;
+      font-size: 10pt;
+      color: gray;
+    }
+  }
```

Comment:
These styles are unnecessarily nested. I usually advise devs to avoid nesting CSS selectors unless there's a very good reason. Among other issues, nesting:

* introduces coupling between DOM structure and style definitions
* increases selector specificity
* increases the size of the actual outputted CSS

A couple references:

* http://csswizardry.com/2012/05/keep-your-css-selectors-short/
* http://thesassway.com/beginner/the-inception-rule

In almost all cases, the code can be flattened with no change in functionality:

```css
.storage-alert {
  font-size: 12pt;
}
.storage-usage {
  margin-top: -5px;
  font-size: 10pt;
  color: gray;
}
```